typed-protocols/src/Network/TypedProtocol/Driver/Limits.hs

```diff
@@ -0,0 +1,189 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeInType #-}
+{-# LANGUAGE BangPatterns #-}
+
+-- | Drivers for running 'Peer's.
+--
+module Network.TypedProtocol.Driver.Limits (
+
+  -- * Limits
+  ProtocolLimits(..),
+  ProtocolLimitFailure(..),
+
+  -- * Normal peers
+  runPeerWithLimits,
+  TraceSendRecv(..),
+
+  -- * Pipelined peers
+  runPipelinedPeerWithLimits,
+
+  -- * Driver utilities
+  driverWithLimits,
+  ) where
+
+import Data.Maybe (fromMaybe)
+
+import Control.Monad.Class.MonadAsync
+import Control.Monad.Class.MonadThrow
+import Control.Monad.Class.MonadTime
+import Control.Monad.Class.MonadTimer
+import Control.Tracer (Tracer (..), traceWith)
+
+import Network.TypedProtocol.Driver.General
+import Network.TypedProtocol.Driver.Simple (TraceSendRecv(..))
+
+import Network.TypedProtocol.Core
+import Network.TypedProtocol.Pipelined
+import Network.TypedProtocol.Channel
+import Network.TypedProtocol.Codec
+
+
+data ProtocolLimits ps bytes = ProtocolLimits {
+       timeLimitForState :: forall (pr :: PeerRole) (st :: ps).
+                            PeerHasAgency pr st -> Maybe DiffTime,
+
+       sizeLimitForState :: forall (pr :: PeerRole) (st :: ps).
+                            PeerHasAgency pr st -> Word,
+
+       dataSize          :: bytes -> Word
+     }
+
+data ProtocolLimitFailure = ExceededSizeLimit
+                          | ExceededTimeLimit
+  deriving Show
+
+instance Exception ProtocolLimitFailure
+
+
+driverWithLimits :: forall ps peerid failure bytes m.
+                    (MonadThrow m, MonadTimer m, Exception failure)
+                 => Tracer m (TraceSendRecv ps peerid failure)
+                 -> peerid
+                 -> Codec ps failure m bytes
+                 -> ProtocolLimits ps bytes
+                 -> Channel m bytes
+                 -> Driver ps (Maybe bytes) m
+driverWithLimits tr peerid Codec{encode, decode}
+                 ProtocolLimits{timeLimitForState, sizeLimitForState, dataSize}
+                 channel@Channel{send} =
+    Driver { sendMessage, recvMessage, startDState = Nothing }
+  where
+    sendMessage :: forall (pr :: PeerRole) (st :: ps) (st' :: ps).
+                   PeerHasAgency pr st
+                -> Message ps st st'
+                -> m ()
+    sendMessage stok msg = do
+      traceWith tr (TraceSendMsg peerid (AnyMessage msg))
```

Comment:
Right, that's what it was here. I reversed this here so that if there's any failure in doing so one could see what it was trying to do, but perhaps consistency is better. I can revert that.

Reply:
It's better to first log and then send consistently :)

Reply:
done